File: LeetCode/Python/Hard/3734. Lexicographically Smallest Palindromic Permutation Greater Than Target/Solution.py
class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        n = len(s)

        # Count characters
        cnt = [0] * 26
        for ch in s:
            cnt[ord(ch) - ord('a')] += 1

        # Check whether a palindrome is possible
        odd = 0
        middle = ""

        for i in range(26):
            if cnt[i] % 2:
                odd += 1
                middle = chr(ord('a') + i)

        if odd > 1:
            return ""

        half_len = n // 2

        # Characters available for first half
        half_cnt = [x // 2 for x in cnt]

        def build(half):
            # First half + middle + reverse(first half)
            return half + middle + half[::-1]

        # Try to match target's first half
        prefix = []
        remaining = half_cnt[:]

        for i in range(half_len):
            c = ord(target[i]) - ord('a')

            if remaining[c] > 0:
                prefix.append(target[i])
                remaining[c] -= 1
            else:
                break

        # If complete first half matches target's first half,
        # check the resulting palindrome.
        if len(prefix) == half_len:
            half = ''.join(prefix)
            candidate = build(half)

            if candidate > target:
                return candidate

        # Try changing a position from right to left.
        cur = prefix[:]
        rem = remaining[:]

        pos = len(cur) - 1

        while pos >= 0:
            current = ord(cur[pos]) - ord('a')

            # Put this character back
            rem[current] += 1

            # Try the smallest character greater than current
            for c in range(current + 1, 26):
                if rem[c] == 0:
                    continue

                half = ''.join(cur[:pos])
                half += chr(ord('a') + c)

                rem[c] -= 1

                # Fill remaining positions with smallest characters
                for x in range(26):
                    half += chr(ord('a') + x) * rem[x]

                candidate = build(half)

                if candidate > target:
                    return candidate

                rem[c] += 1

            pos -= 1
            cur.pop()

        # If no prefix could be matched, try the smallest
        # character greater than target[0].
        if half_len > 0 and len(prefix) == 0:
            first = ord(target[0]) - ord('a')

            for c in range(first + 1, 26):
                if half_cnt[c] == 0:
                    continue

                half = chr(ord('a') + c)

                half_cnt[c] -= 1

                for x in range(26):
                    half += chr(ord('a') + x) * half_cnt[x]

                candidate = build(half)

                if candidate > target:
                    return candidate

                half_cnt[c] += 1

        # n = 1
        if half_len == 0:
            candidate = middle
            if candidate > target:
                return candidate

        return ""

File: LeetCode/C++/Hard/3734. Lexicographically Smallest Palindromic Permutation Greater Than Target/Solution.cpp
class Solution {
public:
    string lexPalindromicPermutation(string s, string target) {
        int n = s.size();

        vector<int> cnt(26, 0);
        for (char c : s)
            cnt[c - 'a']++;

        // A palindrome can have at most one odd frequency.
        int odd = 0;
        char middle = 0;

        for (int i = 0; i < 26; i++) {
            if (cnt[i] % 2) {
                odd++;
                middle = 'a' + i;
            }
        }

        if (odd > 1)
            return "";

        int halfLen = n / 2;

        // Characters available for the first half.
        vector<int> halfCnt(26);
        for (int i = 0; i < 26; i++)
            halfCnt[i] = cnt[i] / 2;

        // Build palindrome from a given first half.
        auto build = [&](string half) {
            string ans = half;

            if (n % 2)
                ans += middle;

            for (int i = halfLen - 1; i >= 0; i--)
                ans += half[i];

            return ans;
        };

        /*
            We construct the first half.

            If we make the first half equal to target's first half,
            the complete palindrome may be > target because of the
            middle/right half.

            Otherwise, we find the first position where our palindrome
            becomes greater and then make the remaining part smallest.
        */

        string prefix;
        vector<int> remaining = halfCnt;

        // Try to match target's first half as long as possible.
        for (int i = 0; i < halfLen; i++) {
            int c = target[i] - 'a';

            if (remaining[c] > 0) {
                prefix += target[i];
                remaining[c]--;
            } else {
                break;
            }
        }

        // Case 1: Entire first half matches target.
        if ((int)prefix.size() == halfLen) {
            string candidate = build(prefix);

            if (candidate > target)
                return candidate;
        }

        /*
            We need to increase some position.

            Start from the rightmost position that can be increased.
            This keeps the resulting string lexicographically smallest.
        */

        string cur = prefix;
        vector<int> rem = remaining;

        int pos = (int)cur.size() - 1;

        while (pos >= 0) {
            int current = cur[pos] - 'a';

            // Restore the character at this position.
            rem[current]++;

            // Try the smallest character greater than current.
            for (int c = current + 1; c < 26; c++) {
                if (rem[c] == 0)
                    continue;

                string half = cur.substr(0, pos);
                half += char('a' + c);

                rem[c]--;

                // Fill the rest with smallest characters.
                for (int x = 0; x < 26; x++) {
                    half += string(rem[x], char('a' + x));
                }

                string candidate = build(half);

                if (candidate > target)
                    return candidate;

                rem[c]++;
            }

            if (pos == 0)
                break;

            // Remove the previous fixed character from cur.
            pos--;
            cur.pop_back();
        }

        /*
            Special case:
            We could not match even one character, so try the smallest
            available character greater than target[0].
        */
        if (halfLen > 0 && cur.empty()) {
            for (int c = target[0] - 'a' + 1; c < 26; c++) {
                if (halfCnt[c] == 0)
                    continue;

                string half;
                half += char('a' + c);

                halfCnt[c]--;

                for (int x = 0; x < 26; x++)
                    half += string(halfCnt[x], char('a' + x));

                string candidate = build(half);

                if (candidate > target)
                    return candidate;

                halfCnt[c]++;
            }
        }

        // n = 1 case
        if (halfLen == 0) {
            string candidate(1, middle);
            if (candidate > target)
                return candidate;
        }

        return "";
    }
};